File: qmdc-py/tests/test_sql.py
"""
Data-driven SQL tests for QMD workspace.

Automatically discovers all directories with tests/ subdirectory containing .sql files.
"""

import json
from pathlib import Path

import pytest

from qmdc.db import QmdcDatabase
from qmdc.workspace import parse_all_workspaces


def find_project_root() -> Path:
    """Find project root by looking for the qmdc-rs/ package directory."""
    current = Path(__file__).resolve()
    while current != current.parent:
        if (current / "qmdc-rs").exists():
            return current
        current = current.parent
    raise RuntimeError("Could not find project root (qmdc-rs/ directory)")


_root = find_project_root()

# Path to scan for test workspaces
SCAN_PATHS = [
    _root / "tests/workspace",
]


def load_workspace_objects(workspace_path: Path) -> list:
    """Load all objects from a workspace (handles multi-workspace folders)."""
    result = parse_all_workspaces(str(workspace_path))
    return result.objects


def get_sql_tests(workspace_path: Path) -> list[tuple[str, Path, Path]]:
    """Find all SQL tests in workspace/tests directory."""
    tests_dir = workspace_path / "tests"
    tests = []

    if not tests_dir.exists():
        return tests

    sql_files = sorted(tests_dir.glob("*.sql"))

    for sql_file in sql_files:
        name = sql_file.stem
        expected_file = tests_dir / f"{name}.expected.json"

        if expected_file.exists():
            tests.append((name, sql_file, expected_file))

    return tests


def find_test_workspaces(directory: Path, prefix: str = "") -> list[tuple[str, Path, Path, Path]]:
    """Recursively find all directories containing tests/ with .sql files."""
    test_cases = []

    if not directory.exists():
        return test_cases

    for entry in sorted(directory.iterdir()):
        if not entry.is_dir():
            continue

        tests_dir = entry / "tests"
        workspace_name = f"{prefix}/{entry.name}" if prefix else entry.name

        # Check if this directory has tests/
        if tests_dir.exists() and tests_dir.is_dir():
            sql_files = list(tests_dir.glob("*.sql"))
            if sql_files:
                for name, sql_file, expected_file in get_sql_tests(entry):
                    test_cases.append((f"{workspace_name}/{name}", entry, sql_file, expected_file))

        # Also check subdirectories (but not tests/ itself)
        if entry.name != "tests":
            test_cases.extend(find_test_workspaces(entry, workspace_name))

    return test_cases


def collect_test_cases():
    """Collect all SQL test cases from all scan paths."""
    test_cases = []

    for scan_path in SCAN_PATHS:
        test_cases.extend(find_test_workspaces(scan_path))

    return sorted(test_cases, key=lambda x: x[0])


TEST_CASES = collect_test_cases()


@pytest.mark.parametrize(
    "test_id,workspace_path,sql_file,expected_file", TEST_CASES, ids=[t[0] for t in TEST_CASES]
)
def test_sql_query(test_id: str, workspace_path: Path, sql_file: Path, expected_file: Path):
    """Run SQL test and compare with expected results."""
    objects = load_workspace_objects(workspace_path)

    db = QmdcDatabase()
    db.sync_objects(objects)

    sql = sql_file.read_text().strip()

    expected = json.loads(expected_file.read_text())
    expected_columns = expected["columns"]
    expected_rows = expected["rows"]

    result = db.query(sql)

    assert result.columns == expected_columns, (
        f"Columns mismatch: expected {expected_columns}, got {result.columns}"
    )

    assert result.rows == expected_rows, (
        f"Rows mismatch:\nexpected: {expected_rows}\ngot: {result.rows}"
    )


def test_parse_all_workspaces():
    """Test that parse_all_workspaces correctly finds all workspaces."""
    test_dir = SCAN_PATHS[0]

    print("\n=== Testing parse_all_workspaces ===")
    print(f"Test directory: {test_dir}")

    result = parse_all_workspaces(str(test_dir))

    workspace_count = sum(1 for obj in result.objects if obj.get("__kind") == "__Workspace")

    print(f"Found {workspace_count} workspace objects")

    assert workspace_count >= 3, f"Expected at least 3 workspaces, found {workspace_count}"

    workspace_ids = [
        obj.get("__id")
        for obj in result.objects
        if obj.get("__kind") == "__Workspace" and obj.get("__id")
    ]

    print(f"Workspace IDs: {workspace_ids}")

    for expected in ["ecommerce", "backend", "frontend"]:
        assert expected in workspace_ids, f"Should find '{expected}' workspace"

    print("✓ parse_all_workspaces test passed")

File: qmdc-ts/tests/test-sql.ts
/**
 * Data-driven SQL tests for QMD workspace.
 *
 * Automatically discovers all directories with tests/ subdirectory containing .sql files.
 */

import { readFileSync, readdirSync, existsSync, statSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { QmdcDatabase } from '../src/db.js';
import { parseAllWorkspaces } from '../src/workspace.js';
import { CaseReport } from './_report.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Path to scan for test workspaces
const SCAN_PATHS = [join(__dirname, '../../tests/workspace')];

interface ExpectedResult {
  columns: string[];
  rows: (string | number | boolean | null)[][];
}

interface SqlTest {
  name: string;
  workspaceName: string;
  workspacePath: string;
  sqlFile: string;
  expectedFile: string;
}

/** Get SQL tests from a workspace directory */
function getSqlTests(workspacePath: string, workspaceName: string): SqlTest[] {
  const testsDir = join(workspacePath, 'tests');
  const tests: SqlTest[] = [];

  if (!existsSync(testsDir)) {
    return tests;
  }

  const sqlFiles = readdirSync(testsDir)
    .filter((f) => f.endsWith('.sql'))
    .sort();

  for (const sqlFile of sqlFiles) {
    const name = sqlFile.replace('.sql', '');
    const expectedFile = join(testsDir, `${name}.expected.json`);

    if (existsSync(expectedFile)) {
      tests.push({
        name,
        workspaceName,
        workspacePath,
        sqlFile: join(testsDir, sqlFile),
        expectedFile,
      });
    }
  }

  return tests;
}

/** Recursively find all directories containing tests/ with .sql files */
function findTestWorkspaces(dir: string, prefix: string = ''): SqlTest[] {
  const tests: SqlTest[] = [];

  if (!existsSync(dir)) {
    return tests;
  }

  const entries = readdirSync(dir);

  for (const entry of entries) {
    const fullPath = join(dir, entry);

    if (!statSync(fullPath).isDirectory()) {
      continue;
    }

    const testsDir = join(fullPath, 'tests');
    const workspaceName = prefix ? `${prefix}/${entry}` : entry;

    // Check if this directory has tests/
    if (existsSync(testsDir) && statSync(testsDir).isDirectory()) {
      const sqlFiles = readdirSync(testsDir).filter((f) => f.endsWith('.sql'));
      if (sqlFiles.length > 0) {
        tests.push(...getSqlTests(fullPath, workspaceName));
      }
    }

    // Also check subdirectories (but not tests/ itself)
    if (entry !== 'tests') {
      tests.push(...findTestWorkspaces(fullPath, workspaceName));
    }
  }

  return tests;
}

/** Collect all tests from all scan paths */
function collectAllTests(): SqlTest[] {
  const tests: SqlTest[] = [];

  for (const scanPath of SCAN_PATHS) {
    tests.push(...findTestWorkspaces(scanPath));
  }

  return tests.sort((a, b) => a.workspaceName.localeCompare(b.workspaceName));
}

async function runTests(report: CaseReport) {
  const allTests = collectAllTests();

  // Group by workspace
  const byWorkspace = new Map<string, SqlTest[]>();
  for (const test of allTests) {
    const key = test.workspaceName;
    if (!byWorkspace.has(key)) {
      byWorkspace.set(key, []);
    }
    byWorkspace.get(key)!.push(test);
  }

  for (const [workspaceName, tests] of byWorkspace) {
    console.log(`\nSQL Tests: ${workspaceName} (${tests.length} tests)`);

    // Load workspace
    const workspacePath = tests[0].workspacePath;
    const result = parseAllWorkspaces(workspacePath);
    const allObjects = result.objects;

    // Create database and sync
    const db = await QmdcDatabase.create();
    db.syncObjects(allObjects);

    for (const test of tests) {
      report.time(`${workspaceName}/${test.name}`, () => {
        const sql = readFileSync(test.sqlFile, 'utf-8').trim();
        const expected: ExpectedResult = JSON.parse(readFileSync(test.expectedFile, 'utf-8'));
        const result = db.query(sql);

        // Compare columns
        if (JSON.stringify(result.columns) !== JSON.stringify(expected.columns)) {
          throw new Error(
            `Columns mismatch: expected ${JSON.stringify(expected.columns)}, got ${JSON.stringify(result.columns)}`
          );
        }

        // Compare rows
        if (JSON.stringify(result.rows) !== JSON.stringify(expected.rows)) {
          throw new Error(
            `Rows mismatch:\nexpected: ${JSON.stringify(expected.rows)}\ngot: ${JSON.stringify(result.rows)}`
          );
        }
      });
    }
  }
}

function testParseAllWorkspaces(report: CaseReport) {
  const artifactsPath = SCAN_PATHS[0];
  console.log('\n=== Testing parseAllWorkspaces ===');
  console.log(`Test directory: ${artifactsPath}`);

  report.time('parse_all_workspaces', () => {
    const result = parseAllWorkspaces(artifactsPath);
    const workspaceCount = result.objects.filter((obj) => obj.__kind === '__Workspace').length;
    console.log(`Found ${workspaceCount} workspace objects`);
    if (workspaceCount < 3) {
      throw new Error(`Expected at least 3 workspaces, found ${workspaceCount}`);
    }
    const workspaceIds = result.objects
      .filter((obj) => obj.__kind === '__Workspace' && obj.__id)
      .map((obj) => obj.__id);
    console.log(`Workspace IDs: ${workspaceIds.join(', ')}`);
    for (const expected of ['ecommerce', 'backend', 'frontend']) {
      if (!workspaceIds.includes(expected)) {
        throw new Error(`Should find '${expected}' workspace`);
      }
    }
  });
}

async function runAllTests() {
  const report = new CaseReport('ts-sql');
  testParseAllWorkspaces(report);
  await runTests(report);
  report.finish();
}

runAllTests().catch((error) => {
  console.error('Test execution failed:', error);
  process.exit(1);
});
